apm-agent-plugins/apm-slf4j-plugin/src/main/java/co/elastic/apm/agent/slf4j/Slf4JMdcActivationListener.java

```diff
@@ -68,37 +78,79 @@ public MethodHandle get(ClassLoader classLoader) {
+            @Nullable
+            @Override
+            public MethodHandle get(ClassLoader classLoader) {
+                try {
+                    return MethodHandles.lookup()
+                        .findStatic(classLoader.loadClass(ORG_SLF4J_MDC), "get", MethodType.methodType(String.class, String.class));
+                } catch (Exception ignore) {
+                    // this class loader does not have the slf4j api
+                    return null;
+                }
+            }
+        });
     @Nullable
     private LoggingConfiguration config;
+    @Nullable
+    private ElasticApmTracer tracer;
 
     @Override
     public void init(ElasticApmTracer tracer) {
+        this.tracer = tracer;
         config = tracer.getConfig(LoggingConfiguration.class);
     }
 
     @Override
     public void onActivate(TraceContextHolder<?> context) throws Throwable {
         if (config != null && config.isLogCorrelationEnabled() && context.isSampled()) {
-            MethodHandle put = mdcPutMethodHandleCache.get(Thread.currentThread().getContextClassLoader());
-            TraceContext traceContext = context.getTraceContext();
+            ClassLoader contextClassLoader = Thread.currentThread().getContextClassLoader();
+
+            // always push something to span ID stack - either valid ID or a non-valid-id marker
+            String formerSpanId = NA_SPAN_ID;
+            MethodHandle get = mdcGetMethodHandleCache.get(contextClassLoader);
+            if (get != null) {
+                formerSpanId = (String) get.invokeExact(SPAN_ID);
+                if (formerSpanId == null) {
+                    formerSpanId = NA_SPAN_ID;
+                }
+            }
+            spanIdStack.get().push(formerSpanId);
+
+            MethodHandle put = mdcPutMethodHandleCache.get(contextClassLoader);
             if (put != null) {
-                put.invokeExact(TRACE_ID, traceContext.getTraceId().toString());
-                String idKey = context instanceof Span ? SPAN_ID : TRANSACTION_ID;
-                put.invokeExact(idKey, traceContext.getId().toString());
+                TraceContext traceContext = context.getTraceContext();
+                put.invokeExact(SPAN_ID, traceContext.getId().toString());
+                if (tracer != null && tracer.isThreadRootContext(context)) {
+                    put.invokeExact(TRACE_ID, traceContext.getTraceId().toString());
+                    put.invokeExact(TRANSACTION_ID, traceContext.getTransactionId().toString());
+                }
             }
         }
     }
 
     @Override
-    public void onDeactivate() throws Throwable {
-        if (config != null && config.isLogCorrelationEnabled()) {
-            MethodHandle remove = mdcRemoveMethodHandleCache.get(Thread.currentThread().getContextClassLoader());
+    public void onDeactivate(TraceContextHolder<?> context) throws Throwable {
```

Comment:
We can't add the `TraceContextHolder` here as it might already be ended and recycled.

Reply:
This method is called from `TraceContextHolder#deactivate`, when the `this` instance is still the active one on the thread stack, so I it is very valid to for checking this state. You are right however that I shouldn't observe it, so I shouldn't call `isSampled` for example.
So, I will add to documentation that I can only use the reference but not observe and certainly not update its state

Reply:
Another option is to transfer on the TraceContextHolder instance hash code...

Reply:
> You are right however that I shouldn't observe it, so I shouldn't call `isSampled` for example.

By providing the context as an argument, we are more likely to make mistakes as that. We should only add the argument if the need for it actually arises. Even then it might more sense to just have the identity hash as a parameter.